fml/raster_thread_merger.h

```diff
@@ -56,6 +60,19 @@ class RasterThreadMerger
   // When task queues are statically merged this method becomes no-op.
   RasterThreadStatus DecrementLease();
 
+  // Record current merge caller in the set of SharedThreadMergerImpl object.
+  // This method should be called before multiple merge callers of same
+  // owner/subsumed pair are going to call |MergeWithLease| method.
+  //
+  // And the reason why not putting this method inside |MergeWithLease| is
+  // |MergeWithLease| should not called when another caller already merged the
+  // threads and |IsMerged| return true. In this case only recording the
+  // current caller is needed.
+  void RecordMergeCaller();
```

Comment:
Feels like we should change the semantics of `MergeWithLease` to be callable by a different thread rather than introduce this rather vague method.

Reply:
@iskakaushik Also MergeWithLease() maybe called on same raster thread with lightweight engine, or maybe called on different threads.  Intruducing this new `RecordMergerCaller()` is marking current caller has platform layers in current frame, and inside `  if (!raster_thread_merger->IsMerged()) ` has some other logics (CancelFrame() return state etc..). See below: 
(in external_view_embedder.cc)

```cpp
PostPrerollResult AndroidExternalViewEmbedder::PostPrerollAction(
    fml::RefPtr<fml::RasterThreadMerger> raster_thread_merger) {
  if (!FrameHasPlatformLayers()) {
    return PostPrerollResult::kSuccess;
  }
  // Record merger firstly when frame has platform layers
  raster_thread_merger->RecordMergeCaller();
  if (!raster_thread_merger->IsMerged()) {
    // The raster thread merger may be disabled if the rasterizer is being
    // created or teared down.
    //
    // In such cases, the current frame is dropped, and a new frame is attempted
    // with the same layer tree.
    //
    // Eventually, the frame is submitted once this method returns `kSuccess`.
    // At that point, the raster tasks are handled on the platform thread.
    raster_thread_merger->MergeWithLease(kDefaultMergedLeaseDuration);
    CancelFrame();
    return PostPrerollResult::kSkipAndRetryFrame;
  }
  raster_thread_merger->ExtendLeaseTo(kDefaultMergedLeaseDuration);
```

Or I may remove `if (!raster_thread_merger->IsMerged())` outside and make the `MergeWithLease` returns a boolean to determine the state?

Reply:
@iskakaushik Also MergeWithLease() maybe called on same raster thread with lightweight engine, or maybe called on different threads.  Intruducing this new `RecordMergerCaller()` is marking current caller has platform layers in current frame, and inside `  if (!raster_thread_merger->IsMerged()) ` has some other logics (CancelFrame() return state etc..). See below: 
(in external_view_embedder.cc)

```cpp
PostPrerollResult AndroidExternalViewEmbedder::PostPrerollAction(
    fml::RefPtr<fml::RasterThreadMerger> raster_thread_merger) {
  if (!FrameHasPlatformLayers()) {
    return PostPrerollResult::kSuccess;
  }
  // Record merger firstly when frame has platform layers
  raster_thread_merger->RecordMergeCaller();
  if (!raster_thread_merger->IsMerged()) {
    // The raster thread merger may be disabled if the rasterizer is being
    // created or teared down.
    //
    // In such cases, the current frame is dropped, and a new frame is attempted
    // with the same layer tree.
    //
    // Eventually, the frame is submitted once this method returns `kSuccess`.
    // At that point, the raster tasks are handled on the platform thread.
    raster_thread_merger->MergeWithLease(kDefaultMergedLeaseDuration);
    CancelFrame();
    return PostPrerollResult::kSkipAndRetryFrame;
  }
  raster_thread_merger->ExtendLeaseTo(kDefaultMergedLeaseDuration);
```

Or I may remove `if (!raster_thread_merger->IsMerged())` outside and make the `MergeWithLease` returns a boolean to determine the state?

Reply:
Is there a way to identify an engine, with an `engine_id` which is unique for each lightweight engine as well? I would like us to consider the tripe: `(engine_id, platform_task_queue_id, raster_task_queue_id)` to determine whether we should extend the lease or Cancel the frame. I think this would also help us avoid statics and storing a set of raw pointers. cc: @gaaclarke

Reply:
@iskakaushik and I talked offline.  I like this idea.  I had a similar comment elsewhere about removing the static variable and putting that information in the Shell.  I think his idea for removing the raw pointers is a good one too.  I think the easiest thing is to make the `engine_id` based on the Engine pointer.  Just don't store it as a pointer so people aren't tempted to call into it.  Something more fancy is probably possible too, but wouldn't be necessary imo.

Reply:
Done! Please review again~ @gaaclarke @iskakaushik